xdsl/dialects/complex.py

```diff
@@ -0,0 +1,381 @@
+    ):
+        lhs_ssa = SSAValue.get(lhs)
+        super().__init__(
+            operands=[lhs_ssa, rhs],
+            result_types=[lhs_ssa.type],
+            properties={"fastmath": fastmath},
+        )
+
+    def verify_(self):
+        if self.lhs.type != self.rhs.type:
+            raise VerifyException("lhs and rhs of binary op must have the same type")
+        if self.lhs.type != self.result.type:
+            raise VerifyException("result type must match operand types")
+
+
+class ComplexCompareOp(IRDLOperation, abc.ABC):
+    """Base class for comparison operations on complex numbers."""
+
+    T: ClassVar = VarConstraint("T", ComplexTypeConstr)
+    lhs = operand_def(T)
+    rhs = operand_def(T)
+    result = result_def(IntegerType(1))
+
+    traits = traits_def(Pure())
+
+    assembly_format = "$lhs `,` $rhs attr-dict `:` type($lhs)"
+
+    def __init__(self, lhs: SSAValue | Operation, rhs: SSAValue | Operation):
+        super().__init__(operands=[lhs, rhs], result_types=[IntegerType(1)])
+
+    def verify_(self):
+        if self.lhs.type != self.rhs.type:
+            raise VerifyException("lhs and rhs of compare op must have the same type")
+
+
+@irdl_op_definition
+class AbsOp(ComplexUnaryRealResultOp):
+    name = "complex.abs"
+
+
+@irdl_op_definition
+class AddOp(ComplexBinaryOp):
+    name = "complex.add"
+
+
+@irdl_op_definition
+class AngleOp(ComplexUnaryRealResultOp):
+    name = "complex.angle"
+
+
+@irdl_op_definition
+class Atan2Op(ComplexBinaryOp):
+    name = "complex.atan2"
+
+
+@irdl_op_definition
+class BitcastOp(IRDLOperation):
+    name = "complex.bitcast"
+    operand = operand_def(AnyAttr())
+    result = result_def(AnyAttr())
+
+    traits = traits_def(Pure())
+
+    assembly_format = "$operand attr-dict `:` type($operand) `to` type($result)"
+
+    def __init__(self, operand: SSAValue | Operation, result_type: Attribute):
+        super().__init__(operands=[operand], result_types=[result_type])
+
+
+@irdl_op_definition
+class ConjOp(ComplexUnaryOp):
+    name = "complex.conj"
+
+
+@irdl_op_definition
+class ConstantOp(IRDLOperation):
+    name = "complex.constant"
+    value = prop_def(ArrayAttr)
+    complex = result_def(ComplexTypeConstr)
+
+    traits = traits_def(Pure(), ConstantLike())
+
+    assembly_format = "$value attr-dict `:` type($complex)"
+
+    def __init__(self, value: ArrayAttr, result_type: ComplexType):
+        super().__init__(properties={"value": value}, result_types=[result_type])
+
+    def verify_(self):
+        if len(self.value.data) != 2:
+            raise VerifyException("complex.constant must have an array of 2 values")
+        elem_type = self.complex.type.element_type
+        # TODO: this check might be wrong, x can be something other than float?
+        if any(cast(FloatAttr, x).type != elem_type for x in self.value.data):
+            raise VerifyException(
+                "complex.constant value types must match the complex element type"
+            )
```

Comment:
I don't yet fully grasp types/constraints in xDSL, but `cast(FloatAttr, x)` is wrong as `x` can be something else.
e.g. `%cst_i1 = complex.constant [1 : i1, 0 : i1] : complex<i1>`

What should the cast be, something related to `ComplexElementCovT`? And probably there needs to be an explicit check as well?

Reply:
I've put comments on this below, but it looks like you should use `ComplexTypeConstr = ComplexType.constr(AnyFloat)` everywhere?